src/main/kotlin/org/rust/ide/typing/paste/RsImportCopyPasteProcessor.kt

```diff
@@ -0,0 +1,257 @@
+        indented: Ref<in Boolean>,
+        values: List<RsTextBlockTransferableData>
+    ) {
+        PsiDocumentManager.getInstance(project).commitAllDocuments()
+
+        val data = values.getOrNull(0) ?: return
+        val rsFile = editor.document.toPsiFile(project) as? RsFile ?: return
+        val range = bounds.range
+
+        val elements = gatherElements(rsFile, range)
+        val importCtx = elements.firstOrNull { it is RsElement } as? RsElement ?: return
+
+        val visitor = ImportingVisitor(importCtx, range, data)
+
+        runWriteAction {
+            for (element in elements) {
+                element.accept(visitor)
+            }
+            // We need to import the candidates after visiting all elements, otherwise the relative offsets could be
+            // invalidated after an import has been added
+            for ((candidate, ctx) in visitor.importCandidates) {
+                candidate.import(ctx)
+            }
+        }
+    }
+
+    companion object {
+        val dataFlavor: DataFlavor? by lazy {
+            try {
+                val dataClass = RsReferenceData::class.java
+                DataFlavor(
+                    DataFlavor.javaJVMLocalObjectMimeType + ";class=" + dataClass.name,
+                    "RsReferenceData",
+                    dataClass.classLoader
+                )
+            } catch (e: NoClassDefFoundError) {
+                null
+            } catch (e: IllegalArgumentException) {
+                null
+            }
+        }
+    }
+}
+
+private class RsReferenceData
+
+private class ImportingVisitor(
+    private val importCtx: RsElement,
+    private val range: TextRange,
+    private val data: RsTextBlockTransferableData
+) : RsRecursiveVisitor() {
+    private val candidates: MutableList<Pair<ImportCandidate, RsElement>> = mutableListOf()
+
+    val importCandidates: List<Pair<ImportCandidate, RsElement>> = candidates
+
+    override fun visitPath(path: RsPath) {
+        val ctx = AutoImportFix.findApplicableContext(path)
+        handleContext(path, ctx)
+        super.visitPath(path)
+    }
+
+    override fun visitMethodCall(methodCall: RsMethodCall) {
+        val ctx = AutoImportFix.findApplicableContext(methodCall)
+        handleContext(methodCall, ctx)
+        super.visitMethodCall(methodCall)
+    }
+
+    override fun visitPatBinding(binding: RsPatBinding) {
+        if (data.importMap.elementToFqn(binding, range) != null) {
+            val referenceName = binding.referenceName
+            val isNameInScope = binding.hasInScope(referenceName, VALUES)
+            if (!isNameInScope) {
+                val importContext = ImportContext2.from(binding, ImportContext2.Type.AUTO_IMPORT)
+                if (importContext != null) {
+                    val candidates = ImportCandidatesCollector2.getImportCandidates(importContext, referenceName)
+                    val ctx = AutoImportFix.Context(AutoImportFix.Type.GENERAL_PATH, candidates)
```

Comment:
This is not a lot of code, but lets wait for #9153 and reuse corresponding function from there